pkg/controllerutils/finalizers.go

```diff
@@ -19,87 +19,37 @@ import (
 
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	"k8s.io/client-go/util/retry"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// PatchAddFinalizers adds the given finalizers to the object via a merge patch request with optimistic locking.
-func PatchAddFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
+// AddFinalizers ensures that the given finalizer is present in the given object with exponential backoff and optimistic
+// locking. If it is not set, it adds it and issues a patch.
+// Note that this is done with a regular merge-patch since strategic merge-patches do not work with custom resources,
+// see https://github.com/kubernetes/kubernetes/issues/105146.
+func AddFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
 	return patchFinalizers(ctx, writer, obj, mergeFromWithOptimisticLock, controllerutil.AddFinalizer, finalizers...)
 }
 
-// PatchRemoveFinalizers removes the given finalizers from the object via a merge patch request with optimistic locking.
-func PatchRemoveFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
-	return patchFinalizers(ctx, writer, obj, mergeFromWithOptimisticLock, controllerutil.RemoveFinalizer, finalizers...)
+// RemoveFinalizers ensures that the given finalizer is not present anymore in the given object with exponential backoff
+// and optimistic locking. If it is set, it removes it and issues a patch.
+// Note that this is done with a regular merge-patch since strategic merge-patches do not work with custom resources,
+// see https://github.com/kubernetes/kubernetes/issues/105146.
+func RemoveFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
+	return client.IgnoreNotFound(patchFinalizers(ctx, writer, obj, mergeFromWithOptimisticLock, controllerutil.RemoveFinalizer, finalizers...))
+}
+
+// RemoveAllFinalizers ensures that the given object has no finalizers with exponential backoff.
+// If any finalizers are set, it removes them and issues a patch.
+func RemoveAllFinalizers(ctx context.Context, writer client.Writer, obj client.Object) error {
+	beforePatch := obj.DeepCopyObject().(client.Object)
+	obj.SetFinalizers(nil)
+	return client.IgnoreNotFound(writer.Patch(ctx, obj, mergeFromWithOptimisticLock(beforePatch)))
```

Comment:
We could even drop optimistic locking here :)
```suggestion
	return client.IgnoreNotFound(writer.Patch(ctx, obj, mergeFrom(beforePatch)))
```